Switch Expander.IsExpanded to a StyledProperty

robloo opened the pull request:
## What does the pull request do?

Switches the Expander.`IsExpanded` property to a `StyledProperty` and adds a new `CancelRoutedEventArgs` based on [`CancelEventArgs`](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.canceleventargs?view=net-7.0).

This is an update to PR https://github.com/AvaloniaUI/Avalonia/pull/9555 based on discussion in #9944

## What is the current behavior?

Expander.`IsExpanded` is a `DirectProperty` and cannot be used with Styling.

## What is the updated/expected behavior with this PR?

 1. Expander.`IsExpanded` is now a full `StyledProperty` and usable in the styling system.
 2. New `CancelRoutedEventArgs` used for both the `Expanding` and `Collapsing` events based on [`CancelEventArgs`](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.canceleventargs?view=net-7.0).

## How was the solution implemented (if it's not obvious)?

See code.

I had to go a different direction than what @tomenscape has done in #9944's working branch. It was necessary to add back some of the original code to:

 1. Call OnExpanding()/OnCollapsing() internally
 3. Notify the property has changed even when it is cancelled. This ensures the control templates toggle button doesn't desync and the animation states are correct.

## Checklist

- ~[ ] Added unit tests (if possible)?~
- [x] Added XML documentation to any related classes?
- ~[ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation~

## Breaking changes

Yes, `StyledProperty` instead of `DirectProperty` but functionality is largely only expanded for app developers.

## Obsoletions / Deprecations

None

## Fixed issues

Relates to #9944

robloo committed:
Switch Expander.IsExpanded to a StyledProperty

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0028643-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo commented on src/Avalonia.Controls/Expander.cs:
```diff
@@ -332,5 +288,63 @@ private void UpdatePseudoClasses()
+                // Remember IsExpanded is usually controlled by a ToggleButton in the control theme and is also used for animations.
+                _ignorePropertyChanged = true;
+
+                RaisePropertyChanged(
+                    IsExpandedProperty,
+                    oldValue: value,
+                    newValue: !value,
+                    BindingPriority.LocalValue,
+                    isEffectiveValue: true);
```
I think everyone agrees that this is non-ideal and kind-of 'hacky'. It was first discussed here: https://github.com/AvaloniaUI/Avalonia/pull/9555#discussion_r1035415374. 

This is why I was hoping coercion would handle this automatically for us but unfortunately it doesn't. Therefore, I had to add this back (you can remove it and test the control catalog to see the issues). I think WPF has issues here as well but for sure it would be better to improve coercion and actually handle the real property value that comes back after coercion itself.

For now, it must be done to ensure external code state is correct. (It is essentially just used to notify of a property changed event and force updates) Keep in mind the expander itself isn't the main 'source' of the IsExpanded property -- it is actually the checked state of a toggle button in the control template... so we have to be careful about this.

> The problem is that when two-way bindings transfer a new value in either direction, they assume that it was accepted by the recipient.

Yes, that's the real issue. One I think exists in WPF as well.

> the new value has local priority 

That's something to think about some more I suppose. I copy pasted the code and when it was a direct-property LocalValue made sense. Now perhaps it doesn't.

grokys replied:
> This is why I was hoping coercion would handle this automatically for us but unfortunately it doesn't.

Yeah I think we should probably make coercion handle this if it's not too difficult. What doesn't work with coercion?

robloo replied:
> What doesn't work with coercion?

Here was a nice summary:

>If the recipient object changes the value to something other than the one that was set, a property changed event is raised and the binding can update the sender object. But if the recipient object instead ignores the new value (as is the case here), then no event is raised and no action is taken.

Basically when a cancel happens in the coerce method the property value is reverted back to its original value. This means the value hasn't actually changed and no change notification happens -- seems to make sense. However, if two controls are two-way bound this actually causes a desync. Because the sender will still change it's state.

My simple understanding (without looking at any control flow) is:
1. Sender changes value (say true -> false)
2. Receiver coerces value
3. Coercion raises cancel events and external code cancels the change (value switch false -> true)
4. The value 'true' is set to the control... but this is the same as it was before. Therefore no change notification and updates.
5. The sender still has a false value instead of true because it never got a change notification because the value didn't actually change.

So an additional check is needed to see if coercion changes the value to a value other than what when into the coerce method. If so, and there is a two-way binding, it needs to notify the sender of a change even when the value set to the control is actually the same.

robloo replied:
True, and the ViewModel case is the issue I've seen in WPF before if I recall correctly. So this is an old problem.

robloo replied:
@grokys  In the example I gave:

 * sender : This is the IsChecked property of the ToggleButton within the control template. 
 * receiver : The actual IsExpanded property of the Expander control itself

This is a general problem though with two-way binding if the "receiver" doesn't accept the sender's value for whatever reason.

grokys replied:
Ok, thanks for the clarification, that makes sense.

The problem with the `RaisePropertyChanged` hack here is that this method is `internal` so the hack will only work within Avalonia itself. We probably need to work out how to do this properly.

robloo replied:
Supporting cancellation in events is a fairly common practice from my experience and the API is mostly standardized as discussed above. It really should be supported in a UI framework IMO. Flyout is another example. If it currently requires an internal-use-only work-around for now that seems justifiable to me (especially a harmless one). The issue with the property system has existed for almost 20 years.

It's common in some cases to need to raise a property changed notification "hack" in apps as well. Especially due to the view model example above. It might also be possible to promote this idea from a "hack" into a "feature" by allowing a clean way to force a refresh.

robloo commented on src/Avalonia.Controls/Expander.cs:
```diff
@@ -332,5 +288,63 @@ private void UpdatePseudoClasses()
+        /// <param name="value">The value to coerce.</param>
+        protected virtual bool OnCoerceIsExpanded(bool value)
+        {
+            RoutedEventArgs eventArgs;
```
Also note I kept using `RoutedEventArgs` here. The other branch created a new `CancelRoutedEventArgs` class with a `Cancel` property (similar to [CancelEventArgs](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.canceleventargs?view=net-7.0)). Personally, I don't think that is needed; however, if others do please speak up.

tomenscape replied:
I added the explicit cancel property on my branch partly so that this and `SplitView` used the same event args, but also because marking the event as handled to cancel the property change is is blurring two separate concepts together.

If a handler sets `e.Handled = true`, then no further event handlers see it (unless the subscribed with `handledEventsToo`, but that is a special case).

Now, what if a control wants to block a parent control from cancelling the change? This isn't possible. Either the control cancels the change itself, or it allows the event to pass on to the parent, which will cancel it.

That scenario won't be common, but it is something that could reasonably happen in someone's `UserControl`.

robloo replied:
As far as I am aware, no `CancelRoutedEventArgs` has ever existed from WPF or in later XAML frameworks. `Handled = true` is used to effectively cancel the event. Yes, this may stop propagation in most cases as well but it hasn't historically been an issue in the core frameworks... that said an internet search will show several control libraries have implemented `CancelRoutedEventArgs` in various places. 

You also make some good points that there are slight differences between marking the event as handled and marking it as cancelled. Maybe uncancelling an event is also something we should support...

With those two points in mind, I'll add the class unless one of the core maintainers disagrees.

robloo replied:
Seems everyone is OK with this and it is preferred so marking resolved.

Linking to other conversions:

 * https://github.com/AvaloniaUI/Avalonia/pull/9555#discussion_r1033128176
 * https://github.com/AvaloniaUI/Avalonia/pull/9555#issuecomment-1383758760

robloo committed:
Add CancelRoutedEventArgs and use it in Expander

maxkatz6 commented on src/Avalonia.Controls/Expander.cs:
```diff
@@ -59,12 +59,11 @@ public class Expander : HeaderedContentControl
-                (o, v) => o.IsExpanded = v,
-                defaultBindingMode: Data.BindingMode.TwoWay);
+                defaultBindingMode: BindingMode.TwoWay,
+                coerce: CoerceIsExpanded);
```
Why not OnPropertyChanged?

robloo replied:
You mean instead of coercion? Because the intention is to intercept the property change BEFORE it happens. We never want to actually change the value if the event is cancelled so we have to raise the events before the property system otherwise handles the change.

maxkatz6 replied:
Makes sense, missed the fact that it is cancellable.

robloo commented:
Tests are failing because this:

```csharp
        public void AddHandler<TEventArgs>(
            RoutedEvent<TEventArgs> routedEvent,
            EventHandler<TEventArgs>? handler,
            RoutingStrategies routes = RoutingStrategies.Direct | RoutingStrategies.Bubble,
            bool handledEventsToo = false) where TEventArgs : RoutedEventArgs
```

Is not matching for the `CancelRoutedEventArgs` handler... which derives directly from `RoutedEventArgs` instead this method is called:

```csharp
        public void AddHandler(
            RoutedEvent routedEvent,
            Delegate handler,
            RoutingStrategies routes = RoutingStrategies.Direct | RoutingStrategies.Bubble,
            bool handledEventsToo = false)
```

This almost seems like a compiler bug to me. If this really doesn't work there is no point in even having that method generic.

grokys commented:
> Is not matching for the CancelRoutedEventArgs handler... which derives directly from RoutedEventArgs instead this method is called:

I think we can probably remove the overload that accepts a `Delegate`?

robloo committed:
Add Interactive.AddHandler() for CancelRoutedEventArgs

This hopefully fixes unit tests due to wrong overload being selected by the compiler.

tomenscape commented:
> This almost seems like a compiler bug to me.

Perhaps it is. I ran all tests locally on the previous commit in your branch, i.e. without `AddHandler`, and they succeeded. I'm using the latest build (17.4.4) of VS 2022. Which test was failing for you?

robloo commented:
> Which test was failing for you?

All tests are failing in the PR (which you might not be able to see) because it won't compile -- it isn't an actual test failure:

![image](https://user-images.githubusercontent.com/17993847/212726357-8923c535-a95e-4d45-95a5-f74cd4be4a19.png)

```
[ERR] Compile: D:\a\1\s\src\Avalonia.Controls\Expander.cs(157,48): error CS8604: Possible null reference argument for parameter 'handler' in 'void Interactive.AddHandler(RoutedEvent routedEvent, Delegate handler, RoutingStrategies routes = RoutingStrategies.Direct | RoutingStrategies.Bubble, bool handledEventsToo = false)'. [D:\a\1\s\src\Avalonia.Controls\Avalonia.Controls.csproj::TargetFramework=net6.0]
[ERR] Compile: D:\a\1\s\src\Avalonia.Controls\Expander.cs(158,54): error CS8604: Possible null reference argument for parameter 'handler' in 'void Interactive.RemoveHandler(RoutedEvent routedEvent, Delegate handler)'. [D:\a\1\s\src\Avalonia.Controls\Avalonia.Controls.csproj::TargetFramework=net6.0]
[ERR] Compile: D:\a\1\s\src\Avalonia.Controls\Expander.cs(166,46): error CS8604: Possible null reference argument for parameter 'handler' in 'void Interactive.AddHandler(RoutedEvent routedEvent, Delegate handler, RoutingStrategies routes = RoutingStrategies.Direct | RoutingStrategies.Bubble, bool handledEventsToo = false)'. [D:\a\1\s\src\Avalonia.Controls\Avalonia.Controls.csproj::TargetFramework=net6.0]
[ERR] Compile: D:\a\1\s\src\Avalonia.Controls\Expander.cs(167,52): error CS8604: Possible null reference argument for parameter 'handler' in 'void Interactive.RemoveHandler(RoutedEvent routedEvent, Delegate handler)'. [D:\a\1\s\src\Avalonia.Controls\Avalonia.Controls.csproj::TargetFramework=net6.0]
```

Setting a breakpoint on the AddHandler methods and running the code confirmed that even on my local computer the wrong method is getting called. It's possible that the build server is set to treat these warnings as errors too which is why it isn't showing up locally (I didn't have a compile problem either).

Unfortunately, even the new `AddHandler` method with explicit `CancelRoutedEventArgs` didn't fix it -- which is very, very strange. Not sure where to go from here and will think about it some more.

tomenscape commented:
I explicitly added generic parameters to the `AddHandler` calls and intellisense revealed the problem: you changed `ExpandedEvent` to `CancelRoutedEventArgs`, when you meant to change `CollapsingEvent`. These two events are on the lines mentioned by the compiler error.

Not sure why error CS8604 is appearing on the CI but not in local builds. Another thing to fix!

robloo commented:
@tomenscape Well, I feel dumb and should have double checked it wasnt something simple. Thanks a lot for taking a look!

robloo committed:
Revert "Add Interactive.AddHandler() for CancelRoutedEventArgs"

This reverts commit cfb90c8201378dab7afc8d75316ea9c069c2c79c.

robloo committed:
Fix expander event args type

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0028774-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into expander-isexpanded-styled-prop

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0028791-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

maxkatz6 commented:
Do I understand it right, the only problem with this PR is RaisePropertyChanged hack? 
If so, can we merge it and move RaisePropertyChanged discussion to the new issue?

robloo commented:
> Do I understand it right, the only problem with this PR is RaisePropertyChanged hack?

Yes

> If so, can we merge it and move RaisePropertyChanged discussion to the new issue?

I've created a new issue: https://github.com/AvaloniaUI/Avalonia/issues/10092.

It will take a while before I could dig into the property system and attempt a fix. Otherwise, I'm probably not the best person to do it (I do have to learn the internals here better someday though). This really does seem harmless for now to me.

maxkatz6 committed:
Merge branch 'master' into expander-isexpanded-styled-prop

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0029150-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]
